tap_facebook/__init__.py

```diff
@@ -278,22 +280,38 @@ def run_job(self, params):
                     'Insights job {} did not start after {} seconds'.format(
                         job_id, INSIGHTS_MAX_WAIT_TO_START_SECONDS))
             elif duration > INSIGHTS_MAX_WAIT_TO_FINISH_SECONDS and status != "Job Completed":
-                pretty_error_message = ('Insights job {} did not complete after {} minutes. ' +
-                    'This is an intermittent error and may resolve itself on subsequent queries to the Facebook API. ' +
-                    'You should deselect fields from the schema that are not necessary, ' +
-                    'as that may help improve the reliability of the Facebook API.')
-                raise Exception(pretty_error_message.format(job_id, INSIGHTS_MAX_WAIT_TO_FINISH_SECONDS//60))
+                base_message = 'Insights job {} did not complete after {} minutes. '.format(job_id,
+                                                                                            INSIGHTS_MAX_WAIT_TO_FINISH_SECONDS//60)
+                if self.can_retry:
+                    LOGGER.info(base_message + 'Retrying...')
+                    raise InsightsJobTimeout()
+                else:
+                    pretty_error_message = (base_message +
+                                            'This is an intermittent error and may resolve itself on subsequent queries to ' +
+                                            'the Facebook API. You should deselect fields from the schema that are not ' +
+                                            'necessary, as that may help improve the reliability of the Facebook API.')
+                    raise Exception(pretty_error_message.format(job_id, INSIGHTS_MAX_WAIT_TO_FINISH_SECONDS//60))
             LOGGER.info("sleeping for %d seconds until job is done", sleep_time)
             time.sleep(sleep_time)
             if sleep_time < INSIGHTS_MAX_ASYNC_SLEEP_SECONDS:
                 sleep_time = 2 * sleep_time
         return job
 
+    def run_job_with_retry(self, params):
```

Comment:
Why write our own retry logic instead of using the backoff decorator we use elsewhere: https://github.com/singer-io/tap-facebook/blob/0d84acfd0506a574f61225f05b6fa975ec0b3d42/tap_facebook/__init__.py#L222-L226

Reply:
Latest commits have a method that's similar, but uses the backoff library to make a second attempt.

Error message is the same for both attempts, but the attempt (or attempts if the backoff pattern is changed) will have a "Retrying Insights job..." log message after the log of the exception message.